### PR TITLE
Add CNN support.

### PR DESCRIPTION
**Pull request by khushi-411**
The PR aims to add CNN support to the model.

Thanks

cc: @khushi-411

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,9 @@ datasets/__pycache__/
 eval/__pycache__/
 models/__pycache__/
 ops/__pycache__/
```
Comment (oke-aditya) [inline .gitignore:5]: Nit:

```suggestion
*.__pycache__
```

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,9 @@ datasets/__pycache__/
...
+
+.github/
+datasets/data/archive/
```
Comment (oke-aditya) [inline .gitignore:8]: Nit: Ignoring the folder itself is enough
```suggestion
datasets/
```

File: train_cnn.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+dataset = datasets.ImageFolder(root = '/home/khushi/Documents/simple-neural-network/datasets/data/archive/natural_images', transform = transform)
+dataloader = torch.utils.data.DataLoader(dataset, batch_size = 32) #, shuffle = True)
+
+cnn = CNN()
+opt = optimizers.Adam(cnn.params, learning_rate=0.001)
+loss = losses.SparseSoftMaxCrossEntropyWithLogits()
+
+for step in range(100):
+    bx, by = dataloader.next_batch()
+    by_ = cnn.forward(bx)
```
Comment (oke-aditya) [inline train_cnn.py:111]: You don't need to use `.forward()` Its called automatically.

File: train_cnn.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    _loss = loss(by_, by)
+    cnn.backward(_loss)
+    opt.step()
+    if step % 5 == 0:
+        ty_ = cnn.forward(train_data)
```
Comment (oke-aditya) [inline train_cnn.py:116]: You would end up doing double forward.

File: train_cnn.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+cnn = CNN()
+opt = optimizers.Adam(cnn.params, learning_rate=0.001)
+loss = losses.SparseSoftMaxCrossEntropyWithLogits()
+
+for step in range(100):
+    bx, by = dataloader.next_batch()
+    by_ = cnn.forward(bx)
+    _loss = loss(by_, by)
+    cnn.backward(_loss)
```
Comment (oke-aditya) [inline train_cnn.py:113]: cnn would keep the track of params. no need to pass _loss
You should rather do
```suggestion
    _loss.backward()
```

File: train_cnn.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+import numpy as np
+import pandas as pd
+
+import torch
+from torch import Tensor
+import torchvision.datasets as datasets
+import torchvision.transforms as transforms
+
+sys.path.insert(1, '/home/khushi/Documents/simple-neural-network/models')
```
Comment (oke-aditya) [inline train_cnn.py:11]: Eh? Why?
Why not declare init files and import like a python package?

**Review by oke-aditya — COMMENTED** [review]
Nice work. I thought I would have a look, and had time to review up a bit

File: .gitignore (hunk 2 above)
Comment (khushi-411) [reply .gitignore:8]: Ah, yes. Thanks (:

File: train_cnn.py (hunk 4 above)
Comment (khushi-411) [reply train_cnn.py:116]: Oops, thanks for pointing it out. Actually, this part of the code is not yet ready. I'm still working on it.

File: train_cnn.py (hunk 6 above)
Comment (oke-aditya) [reply train_cnn.py:11]: > Please don't laugh
Well no. I don't like to demean anyone. Sorry If I hurt you.
Don't worry you can do this into a package. :) Ping me anytime you need help

**Comment by khushi-411** [comment]
Hi, @oke-aditya, I'm very thankful to you and the time you took for reviewing it. Glad to get your viewpoints.😁 I will soon make those enhancements (: 
Thanks!

File: train_cnn.py (hunk 6 above)
Comment (khushi-411) [reply train_cnn.py:11]: > Well no. I don't like to demean anyone. Sorry If I hurt you.

Haha, nothing like that. Maybe it's my inner instincts. LOL

**Comment by khushi-411** [comment]
Closing it. Am not working on it right now :)

**Pull request closed without merging**